src/pure_no_compaction.erl

```diff
@@ -0,0 +1,261 @@
+mutate(Op, VV, {Type, {POLog, Crystal}}) ->
+    {ok, {Type, {orddict:store(VV, Op, POLog), Crystal}}}.
+
+%% @doc Specific query for no polog compaction pure data types.
+-spec query(pure_type:crdt()) -> term().
+query({pure_gcounter, {POLog0, _Crystal}}) ->
+    orddict:fold(
+        fun(_Key, Value, Acc) ->
+            case Value of
+                increment ->
+                    Acc + 1;
+                {increment, Val} ->
+                    Acc + Val
+            end
+        end,
+        0,
+        POLog0
+    );
+query({pure_pncounter, {POLog0, _Crystal}}) ->
+    orddict:fold(
+        fun(_Key, Value, Acc) ->
+            case Value of
+                increment ->
+                    Acc + 1;
+                {increment, Val} ->
+                    Acc + Val;
+                decrement ->
+                    Acc - 1;
+                {decrement, Val} ->
+                    Acc - Val
+            end
+        end,
+        0,
+        POLog0
+    );
+query({pure_gset, {POLog0, _Crystal}}) ->
+    sets:from_list(orddict:fold(
+        fun(_Key, {add, El}, Acc) ->
+                ordsets:add_element(El, Acc)
+        end,
+        ordsets:new(),
+        POLog0
+    ));
+query({pure_twopset, {POLog0, _Crystal}}) ->
+    {Add, Rmv} = orddict:fold(
+        fun(_Key, {Op, El}, {Acc1, Acc2}) ->
+            case Op of
+                add ->
+                    {ordsets:add_element(El, Acc1), Acc2};
+                rmv ->
+                    {Acc1, ordsets:add_element(El, Acc2)}
+            end
+        end,
+        {ordsets:new(), ordsets:new()},
+        POLog0
+    ),
+    sets:from_list(ordsets:subtract(Add, Rmv));
+query({pure_aworset, {POLog0, _Crystal}}) ->
+    POLog1 = orddict:fold(
+        fun(Key0, {Op0, El0}, Acc0) ->
+            orddict:append(El0, {Key0, Op0}, Acc0)
+        end,
+        orddict:new(),
+        POLog0
+    ),
+    sets:from_list(orddict:fold(
+        fun(El1, [_|_], Acc1) ->
+            ElOps = lists:last([L || {El, L} <- orddict:to_list(POLog1), El =:= El1]),
+            case [VV || {VV, Op} <- ElOps, Op =:= add] of
+                [] ->
+                    Acc1;
+                [_|_] ->
+                    case [VV || {VV, Op} <- ElOps, Op =:= rmv] of
+                        [] ->
+                            ordsets:add_element(El1, Acc1);
+                        [_|_] ->
+                            Add = lists:last(lists:usort([VV || {VV, Op} <- ElOps, Op =:= add])),
+                            Rmv = lists:last(lists:usort([VV || {VV, Op} <- ElOps, Op =:= rmv])),
+                            case pure_trcb:happened_before(Add, Rmv) of
+                                true ->
+                                    Acc1;
+                                false ->
+                                    ordsets:add_element(El1, Acc1)
+                            end
+                    end
+            end
+        end,
+        ordsets:new(),
+        POLog1
+    ));
```

Comment:
Reviewed this approach with Georges, and its going to be modified to capture first the sub-set of the polog with only maximal elements and then answer the query from those. This extraction of maximal elements might be doable with generic code, and only the query itself data-type specific.